src/interpolations.jl

```diff
@@ -1264,6 +1250,429 @@ function shape_value(ip::BubbleEnrichedLagrange{RefTriangle, 1}, ξ::Vec{2}, i::
+            product_of = $(name){RefLine,order}(points)
+            ref_coord = Array{Vec{2,Float64},1}(undef,(order+1)^2)
+            for i in 1:order+1, j in 1:order+1
+                ref_coord[i+(j-1)*(order+1)] = Vec(points[i],points[j])
+            end
+            perm = getPermLagrangeQ(order)
+            inv_perm = sortperm(perm)
+            $(name){RefQuadrilateral, order, typeof(product_of), typeof(ref_coord), typeof(perm)}(product_of, ref_coord, perm, inv_perm)
+        end
+        getnbasefunctions(::$(name){RefQuadrilateral,order}) where order = (order + 1)^2
+    end
+end
+
+for (name,  facefuncs) in (
+    (:ArbitraryOrderLagrange,               :facedof_indices),
+    (:ArbitraryOrderDiscontinuousLagrange,  :dirichlet_facedof_indices)
+    )
+    @eval begin
+        function _facedof_interior_indices(::$(name){RefQuadrilateral,order}) where order
+            return (SVector{order-1}((i+4 for i in 1:order-1)),
+                SVector{order-1}((i+order+3 for i in 1:order-1)),
+                SVector{order-1}((i+2*order+2 for i in 1:order-1)),
+                SVector{order-1}((i+3*order+1 for i in 1:order-1)))
+        end
+        function $(facefuncs)(ip::$(name){RefQuadrilateral,order}) where order
+            interior = _facedof_interior_indices(ip)
+            face1 = SVector{order + 1}((i == 1 ? 1 : i == 2 ? 2 : interior[1][i-2] for i in 1:order+1))    
+            face2 = SVector{order + 1}((i == 1 ? 2 : i == 2 ? 3 : interior[2][i-2] for i in 1:order+1))    
+            face3 = SVector{order + 1}((i == 1 ? 3 : i == 2 ? 4 : interior[3][i-2] for i in 1:order+1))    
+            face4 = SVector{order + 1}((i == 1 ? 4 : i == 2 ? 1 : interior[4][i-2] for i in 1:order+1))    
+            return (face1, face2, face3, face4)
+        end
+        function shape_value(ip::$(name){RefQuadrilateral, order}, ξ::Vec{2}, i::Int) where order
+            i > getnbasefunctions(ip) && throw(ArgumentError("no shape function $i for interpolation $ip"))
+            ξ_x = ξ[1]
+            ξ_y = ξ[2]
+            i = ip.perm[i]
+            i_x = (i-1)%(order+1) + 1
+            i_y = (i-1)÷(order+1) + 1
+            ip2 = ip.product_of
+            i_x = ip2.inv_perm[i_x]
+            i_y = ip2.inv_perm[i_y]
+            return shape_value(ip2,Vec(ξ_x),i_x) * shape_value(ip2,Vec(ξ_y),i_y)
+        end
+    end
+end
+
+function celldof_interior_indices(ip::ArbitraryOrderLagrange{RefQuadrilateral,order}) where order
+    ncellintdofs = (order - 1)^2
+    totaldofs = getnbasefunctions(ip)
+    return SVector{ncellintdofs}((totaldofs-ncellintdofs+i for i in 1:ncellintdofs))
+end
+
+##########################
+# Lagrange RefHexahedron #
+##########################
+# Based on vtkTriQuadraticHexahedron (see https://kitware.github.io/vtk-examples/site/Cxx/GeometricObjects/IsoparametricCellsDemo/)
+# Permutation to switch numbering to Ferrite ordering
```

Comment:
Can you elaborate?

Reply:
The first line refers to the numbering convention used (copied from fixed order one)
The second one refers to the permutation that maps from the generated (-1:1)^n coordinates to the ones following the convention (such as {-1,...,1} ->{-1,1,...} in 1D)